### PR TITLE
New Ruby parser

### LINKED ISSUE #120: Ruby parsing issues
- [x] `require 'foobar'`
- [x] `require_relative 'foobar'`
- [x] `include 'foobar'`
- [x] Comments 
- [ ] This line: `tr = zr*zr - zi*zi + cr` in this program fails to parse: http://android-scripting.googlecode.com/svn/trunk/jruby/src/bench/shootout/mandelbrot.ruby
- [ ] This example fails:

``` ruby
def fibonacci( n )
  return  n  if ( 0..1 ).include? n
  ( fibonacci( n - 1 ) + fibonacci( n - 2 ) )
end
puts fibonacci( 5 )
```
- [ ] The opening curly brace in this example causes it to fail:

``` ruby
def fib(n)
  n.times.each_with_object([0,1]) { |num, obj| obj << obj[-2] + obj[-1] }
end
fib(10)
```
- [ ] This one fails:

``` ruby
a = []; a[0] = 1; a[1] = 1
i = 1
while i < 1000000
    a[i+1] = (a[i] + a[i-1])%1000000007
    i += 1
end
puts a[n] 
```
- [ ] A mandelbrot from SO:

``` ruby
puts (-20..20).map{|y|(-40..20).map{|x|z=0;32.times{z=z*z+Complex(x,y)/2e1};z.abs<2??*:' '}*''}
```
- [ ] and one from Rosetta Stone:

``` ruby
require 'complex'

def mandelbrot(a)
  Array.new(50).inject(0) { |z,c| z*z + a }
end

(1.0).step(-1,-0.05) do |y|
  (-2.0).step(0.5,0.0315) do |x|
    print mandelbrot(Complex(x,y)).abs < 2 ? '*' : ' '
  end
  puts
end
```

### PR DESCRIPTION
**Pull request by snim2**
Use an external program to parse Ruby programs.

Fixes #120

### CODE REVIEW

File: lib/eco/rubyparser/rubyparser.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+from export.jruby import JRubyExporter
+from incparser.astree import AST, BOS, EOS, TextNode
+from grammar_parser.gparser import Terminal, Nonterminal
+from incparser.syntaxtable import FinishSymbol
+
+import subprocess
+import tempfile
+
+lexingrules = [
+        ("<ws>", "[ \t]"),
```
Comment (ltratt) [inline lib/eco/rubyparser/rubyparser.py]: Should be a 4 indent not 8?

File: lib/eco/rubyparser/rubyparser.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        ("do", "do"),
+        ("super", "super"),
+        ("self", "self"),
+        ("tIDENTIFIER", "[a-z_][a-zA-Z_0-9]*"),
+        ("tCONSTANT", "[A-Z][a-zA-Z_0-9]*"),
+        ("rest", ".")
+        ]
+
+class RubyProxy(EcoFile):
+
```
Comment (ltratt) [inline lib/eco/rubyparser/rubyparser.py]: Extraneous blank line.

File: lib/eco/treemanager.py [hunk 3]
```diff
@@ -838,6 +838,7 @@ def load_lines(self):
     def save_parsers(self):
         self.saved_parsers[self.version] = list(self.parsers)
 
+
```
Comment (ltratt) [inline lib/eco/treemanager.py]: Blank line?

**Comment by ltratt** [comment]
Modulo some small comments, I like this.

One other comment that doesn't obviously fit anywhere: is the user expected to install ruby-parse somewhere? Could we download it for them automatically or something similar? Or at least offer a config option as to its location (I'm not likely to to want to put it in my `$PATH`!)?

**Comment by snim2** [comment]
In the PR I haven't pushed yet (IYSWIM) I have a rather large change to the Settings dialog (which resolves Issue #133). That will cause a difficult merge conflict if we add the Ruby parser to Settings in this PR, and then merge the other one later on. So, what I propose is:
1. I respond to all your other comments here and we merge this PR
2. I'll cherry-pick the change to Settings from my other branch and push it as a stand-alone PR
3. I'll raise a third PR to add the Ruby parser to the Settings dialog

Does that sound OK?

**Comment by ltratt** [comment]
Sounds like a plan to me (though I'm happy for you to wait until you've pushed the other PR rather than chery-picking stuff out in step 2, if that's a bit awkward for you).

**Comment by snim2** [comment]
Ready to squash?

**Comment by ltratt** [comment]
Yep!

**Comment by snim2** [comment]
OK, squash I think....

**Comment by snim2** [comment]
Yes, that will go away when we move the Ruby parser setting into the Settings dialog.

**Comment by ltratt** [comment]
OK, then I can live with it in the interim :)

**Comment by ltratt** [comment]
Once the tests have run, I'll merge.

**Comment by snim2** [comment]
Thanks

**Pull request merged**